torchtune/models/llama3/_tokenizer.py

```diff
@@ -0,0 +1,180 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+import json
+from pathlib import Path
+from typing import Dict, List, Optional, Tuple
+
+from torchtune.data import Message, truncate
+from torchtune.data.tokenizers import TikTokenEncoding, Tokenizer
+
+
+CL100K_PATTERN = r"""(?i:'s|'t|'re|'ve|'m|'ll|'d)|[^\r\n\p{L}\p{N}]?\p{L}+|\p{N}{1,3}| ?[^\s\p{L}\p{N}]+[\r\n]*|\s*[\r\n]+|\s+(?!\S)|\s+"""  # noqa
+
+PAD_ID = 0
+
+
+class Llama3Tokenizer(Tokenizer):
+    """
+    tiktoken tokenizer configured with Llama3 Instruct's special tokens, as described in
+    https://llama.meta.com/docs/model-cards-and-prompt-formats/meta-llama-3
+
+    Args:
+        path (str): Path to pretrained tiktoken tokenizer file.
+    """
+
+    def __init__(
+        self,
+        path: str,
+    ):
+        all_special_tokens_with_ids = self._get_all_special_tokens_with_ids()
+        self.tt_model = TikTokenEncoding(
+            path=path,
+            name="llama3_tiktoken",
+            pattern=CL100K_PATTERN,
+            special_tokens=all_special_tokens_with_ids,
+        )
+
+        # Encode BOS and EOS, define pad ID
+        self.bos_id = all_special_tokens_with_ids["<|begin_of_text|>"]
+        self.eos_id = all_special_tokens_with_ids["<|end_of_text|>"]
+        self.pad_id = PAD_ID
+
+        # Encode extra special tokens
+        self.step_id = all_special_tokens_with_ids["<|step_id|>"]
+        self.start_header_id = all_special_tokens_with_ids["<|start_header_id|>"]
+        self.end_header_id = all_special_tokens_with_ids["<|end_header_id|>"]
+        self.eom_id = all_special_tokens_with_ids["<|eom_id|>"]
+        self.eot_id = all_special_tokens_with_ids["<|eot_id|>"]
+        self.python_tag = all_special_tokens_with_ids["<|python_tag|>"]
+
+        # During generation, stop when either eos_id or eot_id is encountered
+        self.stop_tokens = [self.eos_id, self.eot_id]
+
+    def _get_all_special_tokens_with_ids(self) -> Dict[str, int]:
```

Comment:
I would prefer this to be configurable with a `file_path` param rather than hardcoded in.